core/arch/arm/kernel/thread.c

```diff
@@ -1645,19 +1645,27 @@ static struct mobj *get_rpc_alloc_res(struct optee_msg_arg *arg,
 	if (arg->ret || arg->num_params != 1)
 		return NULL;
 
+	cookie = arg->params[0].u.tmem.shm_ref;
+
 	psize = READ_ONCE(arg->params[0].u.tmem.size);
+	/*
+	 * There are at least two cases why this is possible:
+	 * - bug in client implementation
+	 * - hypervisor was unable to translate buffer addresses
+	 *   and wants us to call thread_rpc_free()
+	 * In any case, we got buffer that is smaller than expected,
+	 * so we need to free it.
+	 */
 	if (psize < size)
-		return NULL;
+		goto err;
```

Comment:
I think we need to verify that the attribute type is `OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT` before we do anything but treating the reply as garbage.